smithy-typescript-codegen/src/main/resources/software/amazon/smithy/typescript/codegen/integration/http-api-key-auth.ts

```diff
@@ -0,0 +1,122 @@
+// derived from https://github.com/aws/aws-sdk-js-v3/blob/b0e1422630/packages/middleware-host-header/src/index.ts
+
+import { HttpRequest } from "@aws-sdk/protocol-http";
+import {
+  AbsoluteLocation,
+  BuildHandlerOptions,
+  BuildMiddleware,
+  Pluggable,
+} from "@aws-sdk/types";
+
+interface PluginOptions {
+  /**
+   * Where to put the API key.
+   *
+   * If the value is `header`, the API key will be transported in the named header,
+   * optionally prefixed with the provided scheme.
+   *
+   * If the value is `query`, the API key will be transported in the named query parameter.
+   */
+  in: "header" | "query";
+
+  /**
+   * The name of the header / query parameter to use for the transporting the API key.
+   */
+  name: string;
+
+  /**
+   * The scheme to use. Only supported when `in` is `header`.
+   */
+  scheme?: string;
+}
+
+export interface HttpApiKeyAuthInputConfig {
+  /**
+   * The API key to use when making requests.
+   *
+   * This is optional because some operations may not require an API key.
+   */
+  apiKey?: string;
+}
+
+interface PreviouslyResolved {}
+
+export interface HttpApiKeyAuthResolvedConfig {
+  /**
+   * The API key to use when making requests.
+   *
+   * This is optional because some operations may not require an API key.
+   */
+  apiKey?: string;
+}
+
+export function resolveHttpApiKeyAuthConfig<T>(
+  input: T & PreviouslyResolved & HttpApiKeyAuthInputConfig
+): T & HttpApiKeyAuthResolvedConfig {
+  return input;
+}
```

Comment:
OK, thanks... I assume that I also must delete https://github.com/awslabs/smithy-typescript/pull/473/files#diff-d568eb2f093fdfe09e49d3987c65f2a8022f43509666d8b4e3a00d8bd22ecdf6R67-R74 and will need to remove `HttpApiKeyAuthResolvedConfig` interface (etc etc) otherwise bad things will happen.

Reply:
Tried removing `resolvedConfig` and `resolveFunction` as suggested and got the following error:

```
Projection source failed: java.lang.IllegalStateException: If any of inputConfig, resolvedConfig, or resolveFunction are set, then all of inputConfig, resolvedConfig, and resolveFunction must be set: inputConfig: SymbolReference{symbol=./src/middleware/HttpApiKeyAuth/HttpApiKeyAuthInputConfig, alias='HttpApiKeyAuthInputConfig', options=[DECLARE, USE]}, resolvedConfig: null, resolveFunction: null
software.amazon.smithy.typescript.codegen.integration.RuntimeClientPlugin.<init>(RuntimeClientPlugin.java:72)
...
```

Thoughts?

Reply:
Restored `resolvedConfig` and `resolveFunction` (and the resolve function & related types) and things are working again.

Reply:
You can just add pluginFunction and skip inputConfig/resolvedConfig/resolveFunction.

This falls under `HAS_MIDDLEWARE` Convention
https://github.com/awslabs/smithy-typescript/blob/d263078b81485a6a2013d243639c0c680343ff47/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/RuntimeClientPlugin.java#L753-L768

Reply:
Posted suggestion for codegen changes in comment https://github.com/awslabs/smithy-typescript/pull/473#discussion_r774716937

Reply:
I tried this and I get the following compile error when compiling the generated TypeScript:

```
src/AuditClient.ts:203:11 - error TS2345: Argument of type '{ runtime: string; base64Decoder: Decoder; base64Encoder: Encoder; bodyLengthChecker: (body: any) => number | undefined; defaultUserAgentProvider: Provider<UserAgent>; ... 17 more ...; customUserAgent?: string | ... 1 more ... | undefined; } & CustomEndpointsResolvedConfig & RetryResolvedConfig & HostHeaderResolvedC...' is not assignable to parameter of type 'AuditClientResolvedConfig'.
  Type '{ runtime: string; base64Decoder: Decoder; base64Encoder: Encoder; bodyLengthChecker: (body: any) => number | undefined; defaultUserAgentProvider: Provider<UserAgent>; ... 17 more ...; customUserAgent?: string | ... 1 more ... | undefined; } & CustomEndpointsResolvedConfig & RetryResolvedConfig & HostHeaderResolvedC...' is not assignable to type 'Required<ClientDefaults>'.
    Types of property 'apiKey' are incompatible.
      Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.
```

I'm not sure, but this seems to be the problem:

```
type AuditClientResolvedConfigType = __SmithyResolvedConfiguration<__HttpHandlerOptions>
  & Required<ClientDefaults>
```

because it's making the `apiKey` field required instead of optional. I don't know what generates that code or why it's requiring all of the `ClientDefaults` fields to be present.

I think this is what took me down the path of a separate config type and not using `addConfigInterfaceFields` in the first place.

Reply:
@trivikr happy new year, hope you had a chance to take some holidays and get refreshed! Any thoughts on a path forward here? I'm a bit stuck as I can't make the change you requested because the resulting generated code won't compile.

Reply:
@glb Thank you for your patience. The PR is now merged.